### PR TITLE
Implement forecast decomposition for linear models

### LINKED ISSUE #1145: Implement forecast decomposition for linear models
### 🚀 Feature Request

Implement forecast decomposition for `Linear***Model` 

### Proposal

1. In `_SklearnAdapter` create method, which throws `NotImplementedError`:
```
def predict_componetns(df: pd.DataFrame) -> pd.DataFrame
```
2. In `linear.py`:
- Create class `_LinearSklearnAdapter(_SklearnAdapter)`
- Implement `predict_components` method
- Make linear models inherit from `PerSegmentModelMixin`, `NonPredictionIntervalContextIgnorantModelMixin`, `NonPredictionIntervalContextIgnorantAbstractModel`
- Pass the `_LinearSklearnAdapter` in the constructor

3. Implementation ditails:
- Use `feature_names_in_` attribute if linear model to get the names of components
- One of the components is `intercept_`, check that it exists
- To get the component values do the matrix multiplication of features in `df` and `coef_` from the linear model

### Test cases
For models from `linear.py`
- Test components sum up to target
- Test that the components names are correct for couple of feature sets 
- Test that depending on `fit_intercept` parameter we get the correct number of components

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by alex-hse-repository**
## Before submitting (must do checklist)

- [ ] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs? We use Numpy format for all the methods and classes.
- [ ] Did you write any new necessary tests?
- [ ] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Closing issues

closes #1145

### CODE REVIEW

File: etna/models/linear.py [hunk 1]
```diff
@@ -1,11 +1,51 @@
...
+        component_names = [f"target_component_{component_name}" for component_name in self.regressor_columns]
+        target_components = df[self.regressor_columns]
+        components_coefs = self.model.coef_
+
+        if self.model.fit_intercept:
+            component_names.append("target_component_intercept")
+            target_components["intercept"] = 1
+            components_coefs = np.hstack((components_coefs, [self.model.intercept_]))
+
+        target_components = components_coefs * target_components.values
```
Comment (brsnw250) [inline etna/models/linear.py:39]: Here we can directly multiply target components dataframe by model parameters and don't invoke `pd.DataFrame`  constructor.

**Review by brsnw250 — CHANGES_REQUESTED** [review]

File: etna/models/linear.py [hunk 2]
```diff
@@ -1,11 +1,51 @@
...
+        -------
+        :
+            dataframe with prediction components
+        """
+        if self.regressor_columns is None:
+            raise ValueError("Model is not fitted! Fit the model before estimating forecast components!")
+
+        component_names = [f"target_component_{component_name}" for component_name in self.regressor_columns]
+        target_components = df[self.regressor_columns]
+        components_coefs = self.model.coef_
```
Comment (Mr-Geekman) [inline etna/models/linear.py:32]: Are we really sure that ordering of `self.model.coef_` is aligned with ordering of `self.regressor_columns`?

Comment (Mr-Geekman) [reply etna/models/linear.py:32]: May be we can extract the exact names of columns and use them for alignment.

File: etna/models/linear.py [hunk 3]
```diff
@@ -1,11 +1,51 @@
...
-from etna.models.sklearn import SklearnPerSegmentModel
+from etna.models.base import NonPredictionIntervalContextIgnorantAbstractModel
+from etna.models.mixins import MultiSegmentModelMixin
+from etna.models.mixins import NonPredictionIntervalContextIgnorantModelMixin
+from etna.models.mixins import PerSegmentModelMixin
+from etna.models.sklearn import _SklearnAdapter
 
 
-class LinearPerSegmentModel(SklearnPerSegmentModel):
+class _LinearAdapter(_SklearnAdapter):
```
Comment (Mr-Geekman) [inline etna/models/linear.py:13]: Have we wrote anywhere how exactly components are created for linear models? How the user should guess about it?

**Review by Mr-Geekman — CHANGES_REQUESTED** [review]

File: tests/test_models/test_linear_model.py [hunk 4]
```diff
@@ -269,3 +281,47 @@ def test_save_load(model, example_tsds):
...
+    df, regressors = df_with_regressors
+    adapter = _LinearAdapter(regressor=regressor_constructor(fit_intercept=fit_intercept))
+    adapter.fit(df=df, regressors=regressors)
+    target_components = adapter.predict_components(df)
+    assert sorted(target_components.columns) == sorted(expected_component_names)
+
+
+@pytest.mark.parametrize("fit_intercept", (True, False))
+@pytest.mark.parametrize("regressor_constructor", (LinearRegression, ElasticNet))
+def test_linear_adapter_predict_components_sum_up_to_target(df_with_regressors, regressor_constructor, fit_intercept):
```
[the remark this answers is not in the document]
Comment (alex-hse-repository) [reply tests/test_models/test_linear_model.py:321]: In-sample and out-sample mechanics here are the same as the model takes the dataframe with regressors

File: etna/models/linear.py (hunk 3 above)
Comment (alex-hse-repository) [reply etna/models/linear.py:13]: No we haven't))
We haven't done it for any of the models, may be it is better do add notes about it in the models docstring in the separate issue

Comment (alex-hse-repository) [reply etna/models/linear.py:13]: Add Notes here

File: etna/models/linear.py [hunk 5]
```diff
@@ -1,12 +1,56 @@
...
+
+        Returns
+        -------
+        :
+            dataframe with prediction components
+        """
+        if self.regressor_columns is None:
+            raise ValueError("Model is not fitted! Fit the model before estimating forecast components!")
+
+        target_components = df[self.model.feature_names_in_].apply(pd.to_numeric)
```
Comment (brsnw250) [inline etna/models/linear.py:30]: Is it garanteed that `df` will have all columns from `feature_names_in_`?

File: etna/models/linear.py [hunk 6]
```diff
@@ -1,12 +1,56 @@
...
+        components_coefs = self.model.coef_
+
+        if self.model.fit_intercept:
```
Comment (brsnw250) [inline etna/models/linear.py:33]: We can simplify this a bit. Just move this section after line 37 and do `target_components["intercept"] = self.model.intercept_` inside if statement.

**Review by Mr-Geekman — DISMISSED** [review]

**Review by brsnw250 — CHANGES_REQUESTED** [review]

**Review by brsnw250 — APPROVED** [review]

**Pull request merged**